### ISSUE TITLE
Question about namespace resolution

### ISSUE THREAD

**Issue by eagletmt** [itamae-kitchen/mitamae#64]
helper.rb

```ruby
module Helper
  def self.help
    :help
  end
end
```

recipe.rb

```ruby
include_recipe 'helper'
Helper.help
```

In this situation, running `mitamae local recipe.rb` results in the following error.

```
% mitamae version
MItamae v1.5.1
% mitamae local recipe.rb
 INFO : Starting MItamae...
trace:
        [13] /home/eagletmt/.ghq/github.com/k0kubun/mitamae/recipe.rb:2:in instance_eval
/home/eagletmt/.ghq/github.com/k0kubun/mitamae/recipe.rb:2:uninitialized constant #<Class:#<MItamae::RecipeContext:0x559001a94700>>::Helper (NameError)
```

I know changing  `module Helper` to `module ::Helper` solves this, but I'd like to know it's intentional or not.

**Comment by k0kubun** [comment]
I've thought it's fixed in mruby 1.3.0 https://github.com/mruby/mruby/issues/3200 and also in mitamae 1.5.0+, but it may not be the case.

**Comment by k0kubun** [comment]
Ah no, maybe different issue https://github.com/mruby/mruby/issues/3191

**Comment by k0kubun** [comment]
Anyway I suspect it's a bug in mruby side and we should report it.

**Comment by eagletmt** [comment]
OK, so you're saying it is bug and isn't intended behavior? (although it's unclear to me for now it's bug of mruby or mitamae)

**Comment by k0kubun** [comment]
yes

**Comment by k0kubun** [comment]
Basically any behavior different from Itamae should be a bug.

**Comment by k0kubun** [comment]
Looks like this is actually compatible with Itamae. Let me close this until this behavior is changed in Itamae.

```
$ ruby -v
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]

$ itamae version
Itamae v1.11.2

$ cat helper.rb
module Helper
  def self.help
    :help
  end
end

$ cat recipe.rb
include_recipe 'helper'
Helper.help

$ itamae local recipe.rb
 INFO : Starting Itamae...
/home/k0kubun/src/github.com/itamae-kitchen/itamae/recipe.rb:2:in `load': uninitialized constant #<Class:#<Itamae::Recipe::EvalContext:0x00005620a3857888>>::Helper (NameError)
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/recipe.rb:56:in `instance_eval'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/recipe.rb:56:in `load'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/runner.rb:51:in `block in load_recipes'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/runner.rb:42:in `each'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/runner.rb:42:in `load_recipes'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/runner.rb:12:in `run'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/cli.rb:137:in `run'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/lib/itamae/cli.rb:45:in `local'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/thor-1.1.0/lib/thor/command.rb:27:in `run'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/thor-1.1.0/lib/thor/invocation.rb:127:in `invoke_command'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/thor-1.1.0/lib/thor.rb:392:in `dispatch'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/thor-1.1.0/lib/thor/base.rb:485:in `start'
        from /home/k0kubun/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/itamae-1.11.2/bin/itamae:4:in `<top (required)>'
        from /home/k0kubun/.rbenv/versions/3.0.0/bin/itamae:23:in `load'
        from /home/k0kubun/.rbenv/versions/3.0.0/bin/itamae:23:in `<main>'
```

**Comment by k0kubun** [comment]
Although I still prefer having it in Itamae first, it seems we can implement it like:

```diff
diff --git a/mrblib/mitamae/recipe.rb b/mrblib/mitamae/recipe.rb
index 0c10e68..3aeba68 100644
--- a/mrblib/mitamae/recipe.rb
+++ b/mrblib/mitamae/recipe.rb
@@ -37,20 +37,22 @@ module MItamae
     def eval_file(path, variables)
       src = File.read(path)
       context = RecipeContext.new(self, variables)
-      InstanceEval.new(src, path, 1, receiver: context).call
+      InstanceEval.new(src, nil, path, 1, receiver: context).call
     end

     # For #instance_eval with TOPLEVEL_BINDING
     class InstanceEval
       def initialize(*args, receiver:)
-        @args = args
+        receiver.instance_variable_set(:@instance_eval_args, args)
         @receiver = receiver
       end

       def call
-        # When we call #instance_eval, we should not have local variables.
+        # When we call #eval, we should not have local variables.
         # Otherwise a recipe may see the local variables by default.
-        @receiver.instance_eval(*@args)
+        @receiver.instance_eval do
+          eval(*@instance_eval_args)
+        end
       end
     end
   end
```

(thanks to https://dev.to/okuramasafumi/eval-ing-twice-to-define-toplevel-module-from-string-33ma)

**Comment by k0kubun** [comment]
Implemented this feature in https://github.com/itamae-kitchen/mitamae/commit/bb2e449941180b7097febd45f2491ae6a10e747c and released it as v1.12.1.
